client/client.go

```diff
@@ -49,25 +53,46 @@ type Client struct {
 
 	ResponseTimeout time.Duration
 
-	ErrorHandler ErrorHandler
+	ErrorHandler    ErrorHandler
+	handleConnClose ConnCloseHandler
+	handleConnOpen  ConnOpenHandler
 }
 
 // Return a client.
-func New(network, addr string) (client *Client, err error) {
-	conn, err := net.Dial(network, addr)
+func NewNetClient(network, addr string) (client *Client, err error) {
 
-	if err != nil {
+	handlerConnOpen := func() (conn net.Conn, err error) {
+		log.Infof("Trying to connect to server %v ...", addr)
+		for num_tries := 1; ; num_tries++ {
+			if num_tries%100 == 0 {
+				log.Warningf("Still trying to connect to server %v, attempt# %v ...", addr, num_tries)
+			}
+			conn, err = net.Dial(network, addr)
+			if err != nil {
+				time.Sleep(500 * time.Millisecond)
+				continue
+			}
+			log.Infof("Connected to server %v on attempt# %v", addr, num_tries)
+			break
+		}
 		return
 	}
 
-	client = NewConnected(conn)
+	client = NewClient(nil, handlerConnOpen)
 
 	return
 }
 
-// Return a new client from an established connection. Largely used for
-// testing, though other use-cases can be imagined.
-func NewConnected(conn net.Conn) (client *Client) {
+/// handler_conn_close: optional
+func NewClient(handler_conn_close ConnCloseHandler,
+	handler_conn_open ConnOpenHandler) (client *Client) {
+
+	conn, err := handler_conn_open()
+	if err != nil {
```

Comment:
Why not implement the timeout/retry and backoff _here_ instead of in the other location?

Reply:
It would have been implemented here if our Client wasn't compatible with non-TCP type of connection. Since seeing the need for net.Pipe I moved reconnect logic into the handler function.